1단계 미션 제출합니다.

kjk7212 opened the pull request:
어렵네요... 사정없이 말씀해주세요!!

parkjoos reviewed (changes requested):
벨 리팩토링 잘 보았습니다. 전에 말씀 드렸던 부분과 의도에 맞게 잘 변경이 된 것 같아요. 몇가지 의견 추가로 드렸는데 한번 적용가능한 부분이 있는지 확인해 주세요

limmmee-bonjour commented on src/main/java/controller/LoginUserController.java:
```diff
@@ -0,0 +1,34 @@
+        }
+        response.sendRedirect("/user/login_failed.html");
+    }
+}
```
넵 안그래도 이부분은 새로운 기능이 추가될때마다 class가 추가되어 고민이 있었는데..
해당 부분으 조금 더 고민해볼게요 :) 코드리뷰 감사합니다! 👍

limmmee-bonjour commented:
코드리뷰 감사합니다 :) 피드백 주신 부분은 좀 더 고민이 필요할 것 같아 이쯤하고 다시 리뷰 요청 드려요 :)

owenoweny committed:
docs: 기능 목록 작성

owenoweny committed:
feat: HTTP 요청 메세지 파싱 기능 구현

owenoweny committed:
feat: 정적 페이지 GET 요청 응답 기능 구현

owenoweny committed:
refactor: URI 역할 분리

owenoweny committed:
feat: css 컨텐츠타입 처리 기능 구현

owenoweny committed:
refactor: Http 응답 클래스 역할 분리

owenoweny committed:
feat: 회원 저장 기능 구현

- ApiHandler: 어노테이션 기반으로 요청 라우팅 기능 구현
- HttpRequest.headers, body: value 자료형을 String으로 변경
-

owenoweny committed:
feat: /user/create 리다이렉트 기능 구현

kjk7212 committed:
refactor: 상수화, 패키지 정리, handler의 interface화

Rok93 commented on src/test/java/webserver/HttpRequestTest.java:
```diff
@@ -6,6 +6,7 @@
 import org.springframework.web.client.RestTemplate;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
```
불필요한 import문은 제거해주세요. 🙏

추가적으로 아래의 두 테스트가 깨지는 것으로 보이는데, 확인해서 수정 부탁드려요. 🙏😃
* `request_resttemplate`  (1단계 구현 전에는 통과하는 테스트였을거에요. 1단계 구현 이후에도 동일하게 동작하도록 구현해주시면 됩니다. 😉)
* `HTTP_헤더_문자열을_HttpRequest_객체로_파싱한다`

kjk7212 replied:
해당 경로와 매핑된 메소드가 없어서 404를 던지는 것이었습니다!
301 상태코드와 함께 index.html로 리다이렉트 시키면서 해결했습니다!

Rok93 commented on src/main/java/http/URI.java:
```diff
@@ -0,0 +1,46 @@
+
+    private final String path;
+    private final Map<String, String> parameters;
+    private final Optional<MIME> extension;
```
[Java Optional 바르게 쓰기 - 5.Optional을 필드로 사용 금지](https://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/) 글을 참고해보시면 좋을 것 같아요. 🙂


또 하나 생각해볼부분은 URI 객체 입장에서 `extension`은 관심사라고 볼 수 있을까요? 🤔

kjk7212 replied:
extension 상태 자체를 제거하면서 MIME이 필요한 순간에 생성해서 사용하게 변경하였습니다!

Rok93 commented on src/main/java/http/URI.java:
```diff
@@ -0,0 +1,46 @@
+        validatePathFormat(path);
+        this.path = path;
+        this.parameters = parameters;
+        this.extension = Optional.empty();
```
주생성자가 이미 정의되어있으니 `this`를 활용해보면 어떨까요? 😃

kjk7212 replied:
extension을 제거하면서 필요없어져서 아예 제거 했습니다.

Rok93 commented on src/main/java/constant/Constant.java:
```diff
@@ -0,0 +1,13 @@
+package constant;
+
+public class Constant {
```
먼저 `Constant`는 객체 생성을 방지해주시면 좋을 것 같네요. 🙂

상수가 사용되는 위치의 클래스 안에 위치하게된다면 조금 더 응집성이 좋아질 것 같아요. 🤔
상수를 관리하는 위치와 사용하는 위치가 물리적으로 떨어져있다보니 코드의 가독성이 많이 내려가는 것 같아요. 🙂

IDE 환경이라면 상수가 어떤 값인지 확인이라도 할 수 있을텐데 지금과 같이 PR을 보는 상황에서는 더더욱 코드를 파악하기가 어려울 수 있을 것 같아요. 😃

kjk7212 replied:
private 생성자를 만들어 외부에서의 인스턴스 생성을 막았습니다!

확실히 상수는 사용하는 곳이 한 곳일때는 확실히 해당 클래스에 선언하는 것이 좋을 것 같습니다.
그런데 사용하는 클래스가 여러 곳일때는 한 곳에 모아두는 것이 더 좋다는 생각이 드는데 어떻게 생각하시나요??

Rok93 replied:
> 그런데 사용하는 클래스가 여러 곳일때는 한 곳에 모아두는 것이 더 좋다는 생각이 드는데 어떻게 생각하시나요??

사용하는 상수가 여러 곳이라는 의미신거죠? 
딱 적절한 케이스가 떠오르지는 않지만 말씀하신 것처럼 도메인 전반적으로 상수로써 고정되는 값이 있다면 지금같이 `Constant`클래스 같은 곳에 모아두는 것도 나쁘지 않다고 생각해요. 🙂

여러군데에서 쓰이는 상수라면 그 또한 도메인 모델일 수 있다고 생각해요. 
상수들의 공통 요소를 뽑아 enum class로 추출한다거나 할 수도 있으니까요. 😉

Rok93 commented on src/main/java/http/StatusCode.java:
```diff
@@ -0,0 +1,36 @@
+package http;
+
+public enum StatusCode {
```
사소하지만 Http 통신에 대한 StatusCode이니, `HttpStatus`로 명명하는 것이 조금 더 의도를 잘 드러낼 수 있을 것 같네요. 🙂

Rok93 commented on src/main/java/http/HttpRequest.java:
```diff
@@ -0,0 +1,87 @@
+public class HttpRequest {
+    private final URI uri;
+    private final HttpMethod httpMethod;
+    private final String protocol;
```
`uri`, `httpMethod`, `protocol`을 인스턴스 변수로가지는 `RequestLine` 객체를 두고 HttpRequest의 인스턴스 변수로 두면 어떨까요? 🙂

kjk7212 replied:
객체 지향스럽게 HttpRequest가 HttpRequestLine, HttpHeader, HttpBody를 변수로 갖게 변경 했습니다!

Rok93 commented on src/main/java/http/MIME.java:
```diff
@@ -0,0 +1,26 @@
+package http;
+
+public enum MIME {
```
MIME의 각 요소들은 `파일의 확장자`가 하나씩 맵핑되는 것 같은데, 이 확장자들을 MIME 객체의 인스턴스 변수로 두면 어떨까요? 😃
(추가적으로 지금은 URI 객체에있는 불필요한 extension 상태 값을 제거할 수 있을 것 같아요 😃)

kjk7212 replied:
확장자를 추가하면서 해당 변수를 이용해 필요한 순간에만 MIME을 생성하게 변경 했습니다!

Rok93 replied:
아.. 음 클래스 네이밍은   CamelCase를 유지해야하기 때문에 `Mime` 혹은 이게 싫다면 `MimeType` 같은 네이밍으로 변경 부탁드릴게요. 🙏

Rok93 commented on src/main/java/hadler/APIHandler.java:
```diff
@@ -0,0 +1,50 @@
+        try{
+            return (HttpResponse) method.invoke(this, httpRequest);
+        } catch (Exception e) {
+            System.out.println(e);
```
`exception.getMessage()`를 출력하도록 변경해주세요. 🙂

Rok93 commented on src/main/java/hadler/RequestHandler.java:
```diff
@@ -0,0 +1,45 @@
+    private static final ResourceHandler resourceHandler = new ResourceHandler();
+
+    private final Socket connection;
+    private Handler handler = resourceHandler;
```
`handler`를 인스턴스 변수로 두신 이유가 있을까요? 🤔
run 메서드 내부에서 지역변수로 두고 사용해도 충분하지 않을까 싶어서요. 😃

Rok93 commented on src/main/java/hadler/RequestHandler.java:
```diff
@@ -0,0 +1,45 @@
+        logger.debug("New Client Connect! Connected IP : {}, Port : {}", connection.getInetAddress(), connection.getPort());
+
+        try (InputStream in = connection.getInputStream(); OutputStream out = connection.getOutputStream()) {
+            HttpRequest httpRequest = HttpBodyParser.HttpHeaderParsingUtils.parse(new BufferedReader(new InputStreamReader(in)));
```
필요한 위치에서 `BufferedReader` 객체를 생성해서 쓰도록하는게 어떨까 싶어요. 🤔

```suggestion
            HttpRequest httpRequest = HttpBodyParser.HttpHeaderParsingUtils.parse(in);
```

Rok93 commented on src/main/java/hadler/RequestHandler.java:
```diff
@@ -0,0 +1,45 @@
+            }
+
+            HttpResponse httpResponse = handler.handle(httpRequest);
+            httpResponse.send(new DataOutputStream(out));
```
`httpResponse`가 http 응답을 직접 전송하기보다는 httpResponse는 `Response Message`를 만들어주고 실제로 전송하는 부분은 RequestHandler가 수행하도록하면 어떨까요? 🙂

Http 응답을 전송하는 역할은 RequestHandler의 역할로 보여서요. 🙂

kjk7212 replied:
HttpResponse 객체는 바이트로 된 메세지를 생성하고 RequestHandler가 메세지를 전송하게 객체의 역할을 분리했습니다!

Rok93 commented on src/main/java/hadler/ResourceHandler.java:
```diff
@@ -0,0 +1,28 @@
+            System.out.println(e);
+        }
+
+        return HttpResponse.staticResource(body, mime);
```
아래와 같이 변경해보면 조금 더 깔끔해질 것 같아요. 🙂

```suggestion
        try {
            byte[] body = FileIoUtils.loadFileFromClasspath(path + httpRequest.getUri().getPath());
            return HttpResponse.staticResource(body, mime);
        } catch (IOException | URISyntaxException e) {
            System.out.println(e.getMessage());
            throw NotFoundException(); // 해당 예외를 HttpStatusCode가 404인 응답을 HttpResponse를 반환
        }

        return HttpResponse.staticResource(body, mime);
```

Rok93 commented on src/main/java/http/HttpResponse.java:
```diff
@@ -0,0 +1,56 @@
+    private final byte[] body;
+
+    public static HttpResponse found(String redirectURI) {
+        //302일 때 응답 헤더에 필요한 값은??
```
불필요한 주석은 제거해주세요. 🙏

혹시나 해결이 아직 안된 내용이라면 [이 글](https://en.wikipedia.org/wiki/HTTP_302)을 참고해보시면 좋을 것 같아요. 🙂

Rok93 commented on src/main/java/http/HttpResponse.java:
```diff
@@ -0,0 +1,56 @@
+        this.statusCode = statusCode;
+        this.body = body;
+        this.headers = headers;
+    }
```
`생성자` 다음에 `정적 팩토리 메서드` 순서로 위치시켜주세요. 🙏

Rok93 commented on src/main/java/http/HttpResponse.java:
```diff
@@ -0,0 +1,56 @@
+            //TODO: 예외처리
+            System.out.println(e.getMessage());
+        }
+    }
```
마땅히 어떤 예외처리를 해야할지 모르겠어도 일단은 예외를 던지도록 해주는 것이 좋을 것 같아요. 🙂
(어떤의미로는 예상치 못한 에러라고 볼 수 있으니) InternalServerError를 응답해주면 좋을 것 같습니다. 🙂

Rok93 commented on src/main/java/utils/HttpBodyParser.java:
```diff
@@ -0,0 +1,186 @@
+            return new URI(path, parameters);
+        }
+
+        public static Map<String, String> parseParameters(String stringUri) {
```
쿼리 파라미터에 대해서 decoding이 필요할 것 같아요. 🙂

<img width="1005" alt="image" src="https://github.com/next-step/jwp-was-kakao/assets/47850258/e7813e6e-3b3e-4584-b5c4-afdb71a1a2d5">

kjk7212 replied:
까먹었습니다... URLDecoder를 통해 decode하는 부분을 추가했습니다!

Rok93 commented on src/main/java/utils/HttpBodyParser.java:
```diff
@@ -0,0 +1,186 @@
+
+    public static class HttpHeaderParsingUtils {
+        public static HttpRequest parse(BufferedReader bufferedReader) throws IOException {
+            String[] f = bufferedReader.readLine().split(SPACE);
```
네이밍만 봤을 때, 어떤 의도인지 정확히 파악하기가 힘든 것 같아요. 🥲
조금 더 의도를 드러낼 수 있는 네이밍으로 변경해보면 좋을 것 같아요.

kjk7212 replied:
해당 내부 클래스를 제거하고 네이밍을 신경썼습니다!

Rok93 commented on src/main/java/mapper/APIMapper.java:
```diff
@@ -0,0 +1,30 @@
+import webserver.URIHandlerMapping;
+
+public class APIMapper {
+	public static final Map<HttpMethod, URIHandlerMapping> APIMappings;
```
전역변수를 외부에 직접 노출하기보다는, 메서드를 통해서 APIMapper 객체에게 요청을 보내도록 변경해보면 어떨까요?

kjk7212 replied:
APIMappings 변수를 private 처리하고 APIMapper 객체의 메소드를 호출해 존재 여부를 판단하고 가져오게 변경 했습니다!

Rok93 commented on src/main/java/utils/FormUrlEncodedParsingStrategy.java:
```diff
@@ -0,0 +1,28 @@
+    @Override
+    public Map<String, String> parse(String body) {
+        if (body.isBlank()) {
+            return Map.of();
```
아래와 같이 비어있는 Collection에 대해서는 `Collections.emptyMap();`를 활용하는 것이 좋을 것 같아요. 🙂
```suggestion
            Collections.emptyMap();
```

Rok93 commented on src/main/java/utils/HttpBodyParser.java:
```diff
@@ -0,0 +1,186 @@
+                    .build();
+        }
+
+        public static HttpRequest parse(String httpRequestString) {
```
테스트 클래스에서만 사용하는 메서드로 보이네요.
별도로 사용하는 위치가 없다면 과감히 제거해주세요, 🙏

kjk7212 replied:
리팩토링 과정에서 기존것을 남겨 뒀던 것 같습니다... 제거 했습니다!

Rok93 commented on src/main/java/utils/FormUrlEncodedParsingStrategy.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class FormUrlEncodedParsingStrategy implements HttpBodyParsingStrategy {
```
성국님이 `utils` 패키지에 생성하신 Util 클래스가 3개 있는데요.

`FileIoUtils`와 `IOUtils`와 같이 정말로 공용적으로 쓰일 수 있는 유틸리티 성격인지 고민해보시면 좋을 것 같아요. 🙂
(utils 패키지에 둘만한 클래스들인지 아닌지 고민해보시면 좋을 것 같습니다. 😉)

kjk7212 replied:
새 패키지를 생성하면서 좀 더 맞는 곳으로 이동시켰습니다!

Rok93 commented on src/main/java/utils/FormUrlEncodedParsingStrategy.java:
```diff
@@ -0,0 +1,28 @@
+            map = Arrays.stream(body.split(PARAMETER_SEPARATOR))
+                    .map(s -> s.split(PARAMETER_EQUAL_SIGN))
+                    .map(arr -> Map.entry(arr[0], arr[1]))
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```
`Map.Entry`로 변경할 필요 없이, 아래와 같이 바로 변경하면 될 것 같아요. 🙂
```suggestion
                    .collect(Collectors.toMap(arr -> arr[0], arr -> arr[1]));
```

Rok93 commented on src/main/java/utils/FormUrlEncodedParsingStrategy.java:
```diff
@@ -0,0 +1,28 @@
+            throw new IllegalArgumentException("유효하지 않은 형식입니다.");
+        }
+
+        return map;
```
try 블록 내부에서 return을 해준다면  try 블록 밖에서 map 변수를 선언할 필요가 없겠네요. 🙂 
(선언과 동시에 할당을 수행할 수 있겠네요. )

그런데 try 블록 안에서 바로 리턴하게된다면 애초에 변수 선언이 불필요할 것 같아요. 😃

```suggestion
        try {
            return Arrays.stream(body.split(PARAMETER_SEPARATOR))
                    .map(s -> s.split(PARAMETER_EQUAL_SIGN))
                    .map(arr -> Map.entry(arr[0], arr[1]))
                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
        } catch (Exception e) {
            throw new IllegalArgumentException("유효하지 않은 형식입니다.");
        }
```

이 부분에서 조금 더 고민해볼 부분은 `PARAMETER_SEPARATOR`로 추출한 각 **요청 파라미터 key, value 라인**이 유효하지 않은 경우에 대해서도 처리해주면 어떨까요? 😃

kjk7212 replied:
조금 고민을 해봤는데, header의 경우는 몇가지 제약사항이 있어 검증하는 로직을 집어넣었지만
body 내부의 값에 대해서는 잘 모르겠습니다. 어떤 제약 조건이 있을까요???

Rok93 replied:
제가 조금 모호하게 말씀을 드린 것 같네요. 🙂
key와 value가 한 쌍을 이루지 않은 케이스는 예외를 던진다거나하는 유효성 검증 로직을 넣어보면 좋을 것 같다는 이야기였어요. 😃

kjk7212 replied:
쌍이 이루어지지 않거나 할 때 split + 직접 접근을 하다보니 일단은 
ArrayIndexOutOfBoundsException이 발생하는 것을 확인 했습니다!
나중에 시간이 남을때 추가적인 검증 로직을 추가하겠습니다!

Rok93 reviewed (changes requested):
성국님 안녕하세요. 웹 서버 구현하기 미션을 함께하게된 김경록입니다. 🙇‍♂️
1단계 구현 잘해주셨습니다. 👍💯
몇몇 코멘트 남겨두었으니 확인해서 반영해주세요. 🙏
미션 진행하면서 어려운 부분이 있으면 DM 주세요. 😃

kjk7212 committed:
feat: extension의 관리 책임을 MIME으로 넘김

kjk7212 committed:
refactor: 초 대규모 리팩토링

Rok93 commented on src/main/java/handler/APIHandler.java:
```diff
@@ -0,0 +1,61 @@
+	private static final APIMapper apiMapper = new APIMapper();
+
+	public HttpResponse handle(HttpRequest httpRequest) {
+		HttpMethod httpMethod = httpRequest.getHttpRequestLine().getHttpMethod();
```
`httpRequest` [객체에게 메시지를 보내](https://tecoble.techcourse.co.kr/post/2020-04-28-ask-instead-of-getter/)보면 좋을 것 같아요.

이 케이스 말고도 getter를 연달아 두번이상 호출하고있다면 동일하게 적용해줄 수 있을 것 같아요. 🙂

kjk7212 replied:
get.get을 제거 했습니다!

Rok93 commented on src/main/java/handler/APIHandler.java:
```diff
@@ -0,0 +1,61 @@
+			return (HttpResponse)method.invoke(this, httpRequest);
+		} catch (Exception e) {
+			System.out.println(e.getMessage());
+			return HttpResponse.internalServerError(EMPTY_ERROR_MESSAGE);
```
사소한 의견이지만 `Internal Server Error` 던질 때, body 값은 그냥 null로 줘도 될 것 같은데, 굳이 빈 문자열을 넘겨줘야하나 싶어요. 🤔  정적 팩토리 메서드의 메서드 파라미터로 아무것도 받지 않아도되지 않을까요?  (HttpResponse 생성할 때는 body를 그냥 null로 넣어주면되지 않을까 싶었어요 🙂)

kjk7212 replied:
생각해보니 Internal Server Error를 던질 때는 body에 값이 필요 없을 것 같네요! 삭제 했습니다!

Rok93 commented on src/main/java/handler/APIHandler.java:
```diff
@@ -0,0 +1,61 @@
+		}
+		DataBase.addUser(new User(userId, password, name, email));
+
+		return HttpResponse.found("http://localhost:8080/index.html");
```
host url은 제외하고 url `path`만 입력해도 동일하게 동작할 것 같아요. 🙂
```suggestion
		return HttpResponse.found("/index.html");
```

Rok93 commented on src/main/java/handler/RequestHandler.java:
```diff
@@ -0,0 +1,51 @@
+			HttpRequest httpRequest = HttpRequest.createHttpRequestFromInputStream(in);
+
+			Handler handler = apiHandler;
+			;
```
불필요한 `;` 같네요. 제거해주세요. 🙏
```suggestion
```

Rok93 commented on src/main/java/http/HttpResponse.java:
```diff
@@ -0,0 +1,81 @@
+	}
+
+	public byte[] makeByteMessage() {
+		StringBuilder header = new StringBuilder(
+			"HTTP/1.1 " + httpStatus.getCode() + " " + httpStatus.getDescription() + " \r\n");
```
httpStatus 객체에게 메시지를 보내서 `"HTTP/1.1 " + httpStatus.getCode() + " " + httpStatus.getDescription() + " \r\n"` 의 응답을 만들어달라고 할 수 있지 않을까요? 🙂

```suggestion
		StringBuilder header = new StringBuilder(httpStatus.xxx());
```

kjk7212 replied:
객체들에게 문자열 생성을 맡기고 밖에서 합치게 수정했습니다!

Rok93 commented on src/main/java/http/HttpResponse.java:
```diff
@@ -0,0 +1,81 @@
+	public byte[] makeByteMessage() {
+		StringBuilder header = new StringBuilder(
+			"HTTP/1.1 " + httpStatus.getCode() + " " + httpStatus.getDescription() + " \r\n");
+		httpHeader.getHeader()
+			.forEach((key, value) -> header.append(makeHeaderToString(key, value)));
+		header.append("\r\n");
```
마찬가지로 여기의 응답도 header 객체에게 만들어달라고 요청해볼 수 있겠네요. 🙂

kjk7212 replied:
객체들에게 문자열 생성을 맡기고 밖에서 합치게 수정했습니다!

Rok93 commented on src/main/java/http/MIME.java:
```diff
@@ -0,0 +1,42 @@
+		return this.equals(MIME.HTML) || this.equals(MIME.TXT);
+	}
+
+	public static MIME findMimeFromExtension(String extension) {
```
`MIME.xxx`와 같이 MIME 이 prefix로 붙은 꼴이되기 때문에  메서드 네이밍에 꼭 enum class 명이들어가지는 않아도 괜찮을 것 같아요. 🙂

```suggestion
	public static MIME findByExtension(String extension) {
```

Rok93 commented on src/main/java/http/MIME.java:
```diff
@@ -0,0 +1,42 @@
+
+	public static MIME findMimeFromExtension(String extension) {
+		return Arrays.stream(values())
+			.filter(mime -> mime.getExtension().equals(extension))
```
같은 클래스 내부에서는 인스터스 변수에 직접 접근할 수 있어요! 🙂

```suggestion
			.filter(mime -> mime.extension.equals(extension))
```

Rok93 commented on src/main/java/mapper/APIMapper.java:
```diff
@@ -0,0 +1,36 @@
+import http.HttpMethod;
+
+public class APIMapper {
+	private static final Map<HttpMethod, APIHandlerMapping> APIMappings;
+
+	static {
+		APIMappings = Arrays.stream(HttpMethod.values())
+			.collect(Collectors.toMap(Function.identity(), httpMethod -> new APIHandlerMapping()));
+
+		Method[] methods = APIHandler.class.getDeclaredMethods();
+
+		Arrays.stream(methods)
+			.filter(method -> method.isAnnotationPresent(APIMapping.class))
+			.map(method -> new AbstractMap.SimpleEntry<>(method.getAnnotation(APIMapping.class), method))
+			.forEach(entry -> APIMappings.get(entry.getKey().httpMethod()).addMapping(entry.getKey().path(), entry.getValue()));
+	}
```
뭔가 좋은 접근인 것 같은데, 코드가 복잡해서 이해하기가 조금 어려운 것 같아요. 🤔😅

개인적인 생각이지만,  `APIMappings`의 key 값을 `path`로 두고 APIHandlerMapping의 key 값을 HttpMethod로 두는게 조금 더 좋은 방향이 되지 않을까 싶어요. 🙂

kjk7212 replied:
좀 더 좋은 방식을 고민해보겠습니다 🤔

Rok93 commented on src/main/java/parser/FormUrlEncodedParsingStrategy.java:
```diff
@@ -0,0 +1,29 @@
+		}
+
+		try {
+			return new HttpRequestBody(Arrays.stream(body.split(PARAMETER_SEPARATOR))
+				.map(s -> s.split(PARAMETER_EQUAL_SIGN))
+				.collect(Collectors.toMap(arr -> URLDecoder.decode(arr[0], StandardCharsets.UTF_8), arr -> URLDecoder.decode(arr[1], StandardCharsets.UTF_8))));
```
아래와 같이 생성자 파라미터에 통째로 들어간 stream 구문을 변수로 추출해주면 조금 더 가독성이 좋아질 것 같아요. 🙂
```suggestion
                         Map<String, String> body = Arrays.stream(body.split(PARAMETER_SEPARATOR))
				.map(s -> s.split(PARAMETER_EQUAL_SIGN))
				.collect(Collectors.toMap(arr -> URLDecoder.decode(arr[0], StandardCharsets.UTF_8), arr -> URLDecoder.decode(arr[1], StandardCharsets.UTF_8)));
			return new HttpRequestBody(body);
```

Rok93 commented on src/main/java/parser/HttpRequestParser.java:
```diff
@@ -0,0 +1,112 @@
+		this.httpBodyParsingStrategy = null;
+	}
+
+	private HttpRequestBody parseHttpBody(String bodyString) {
```
이 메서드는 `parseHttpRequestBody` 메서드에서 쓰이는데, 뜬금없이 맨 위에 위치해서 코드를 파악하는데 흐름을 깨는 느낌이에요. 🤔
parseHttpRequestBody 메서드 다음에 위치시키면 조금 더 코드를 읽기가 수월할 것 같아요. 🙂

Rok93 commented on src/main/java/parser/HttpRequestParser.java:
```diff
@@ -0,0 +1,112 @@
+		throw new IllegalArgumentException("지원되지 않는 컨텐츠타입입니다.");
+	}
+
+	private URI makeURIFromPath(String path) {
+		Map<String, String> parameters = new HashMap<>();
+		if (hasQuery(path)) {
+			int queryStartIndex = path.indexOf(QUERY_SEPARATOR);
+			path = path.substring(0, queryStartIndex);
+			parameters = parseParameters(path);
+		}
+		return new URI(path, parameters);
+	}
```
가급적 변수를 할당함과 동시에 선언하도록해보시면 좋을 것 같아요. 🙂
```suggestion
	private URI makeURIFromPath(String path) {
		if (hasQuery(path)) {
			int queryStartIndex = path.indexOf(QUERY_SEPARATOR);
			path = path.substring(0, queryStartIndex);
			Map<String, String> parameters = parseParameters(path);

			return new URI(path, parameters);
		}
		
		return new URI(path, Collections.emptyMap());
	}
```

Rok93 commented on src/main/java/parser/HttpRequestParser.java:
```diff
@@ -0,0 +1,112 @@
+			return new HttpRequestBody();
+		}
+		if (CONTENT_TYPE_FORM.equals(httpHeader.getValueByKey(HEADER_CONTENT_TYPE))) {
+			this.httpBodyParsingStrategy = new FormUrlEncodedParsingStrategy();
```
일단 null로 뒀다가 필요한 시점에 구현체를 바꿔쓰는 방식은 언제 이 인스턴스 변수가 할당되었는지에 대해 확신할 수 없고 잘못 사용할 여지가 생기게됩니다. (*NPE 발생 확률 🆙)

전략 인터페이스의 구현체가 여러가지라면 이 전략 구현체 중에서 적절한 구현체를 찾아주는 추상계층(예로들면 xxxFactory)을 하나 둔다거나 해볼 수 있을 것 같아요. 🙂

하지만 지금은 전략 구현체가 하나뿐이라 사실 이 추상계층이 크게 필요해보이지는 않네요. 🤔 
또한 Http Protocol이 바뀌지 않는다면 추가되는 전략 구현체도 없을 것 같은데 Http Protocol이 변경될 확률은 적어보여서 사실 그냥 `FormUrlEncodedParsingStrategy` 객체를 미리 할당해놓고 써도 무방하지 않을까 싶어요.

kjk7212 replied:
확실히 안 좋은 방식인것 같습니다. 파싱하는 방식이 바뀔걸 대비하기 위해 해둔 방식인데,
일단은 FormUrlEncodedParsingStrategy를 할당하는 방식으로 바꾸고 todo에 추상계층 추가를 넣어두었습니다!

Rok93 reviewed (approved):
성국님 피드백 반영 잘해주셨습니다. 👍💯
1단계 PR은 이만 머지하도록하겠습니다. 😉
추가로 몇몇 코멘트 남겨두었는데, 확인해서 다음 단계 진행하면서 같이 반영해주시면 될 것 같습니다. 🙏

2단계에서 뵙겠습니다! 😁
